File: lib/impressioner_web/channels/control_channel.ex
defmodule ImpressionerWeb.ControlChannel do
  use Phoenix.Channel

  def join("control", _message, socket) do
    current_slide = Impressioner.SlideSession.get()
    {:ok, %{"slide" => current_slide}, socket}
  end

  def handle_in("show slide", %{"slide" => slide}, socket) do
    broadcast!(socket, "show slide", %{"slide" => slide})
    Impressioner.SlideSession.set(slide)
    {:noreply, socket}
  end
end

File: frontend/application/channels/controlChannel.js
import socket from '../utils/socket';

const controlChannel = socket.channel('control', {});
const join = controlChannel.join();

const showSlide = (slide) => {
  controlChannel.push('show slide', { slide });
};

const onShowSlide = (callback) => {
  join.receive('ok', (response) => { callback(response.slide); });
  controlChannel.on('show slide', data => callback(data.slide));
};

export {
  showSlide,
  onShowSlide,
};
